presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPartitionManager.java

```diff
@@ -59,10 +91,197 @@ public List<String> getEffectivePartitions(
+        log.info(format("Total partition size is %s, after partition prune size is %s.", matchedPartitionPaths.size(), result.size()));
+        return result;
+    }
+
+    /**
+     * Returns the partition path key and values as a list of map.
+     * For example:
+     * partition keys: [p1, p2, p3],
+     * partition paths:
+     *      p1=val1/p2=val2/p3=val3  (hive style partition)
+     *      p1=val4/p2=val5/p3=val6  (hive style partition)
+     * return values {p1=val1/p2=val2/p3=val3 -> {p1 -> val1, p2 -> value2, p3 -> value3}}, {p1=val4/p2=val5/p3=val6 -> {p1 -> val4, p2 -> value5, p3 -> value6}}
+     *
+     * @param partitionKey  The partition key list
+     * @param partitionPaths partition path list
+     */
+    public static Map<String, Map<String, String>> getPartitions(List<String> partitionKey, List<String> partitionPaths)
+    {
+        Map<String, Map<String, String>> result = new HashMap<>();
+        if (partitionPaths.isEmpty() || partitionKey.isEmpty()) {
+            return result;
+        }
+        // try to infer hive style
+        boolean hiveStylePartition = HIVE_PARTITION_NAME_PATTERN.matcher(partitionPaths.get(0).split(Path.SEPARATOR)[0]).matches();
+        for (String partitionPath : partitionPaths) {
+            String[] pathParts = partitionPath.split(Path.SEPARATOR);
+            Map<String, String> partitionMapping = new LinkedHashMap<>();
+            if (hiveStylePartition) {
+                Arrays.stream(pathParts).forEach(p -> {
+                    String[] keyValue = p.split("=");
+                    if (keyValue.length == 2) {
+                        partitionMapping.put(keyValue[0], keyValue[1]);
+                    }
+                });
+            }
+            else {
+                for (int i = 0; i < partitionKey.size(); i++) {
+                    partitionMapping.put(partitionKey.get(i), pathParts[i]);
+                }
+            }
+            result.put(partitionPath, partitionMapping);
+        }
+        return result;
+    }
+
+    private List<String> prunePartitions(
+            TupleDomain<String> partitionPredicate,
+            List<HudiColumnHandle> partitionColumnHandles,
+            Map<String, Map<String, String>> candidatePartitionPaths)
+    {
+        return candidatePartitionPaths.entrySet().stream().filter(f -> {
+            Map<String, String> partitionMapping = f.getValue();
+            return partitionMapping.entrySet().stream().allMatch(p -> evaluatePartitionPredice(partitionPredicate, partitionColumnHandles, p.getValue(), p.getKey()));
+        }).map(entry -> entry.getKey()).collect(Collectors.toList());
+    }
+
+    private boolean evaluatePartitionPredice(TupleDomain<String> partitionPredicate, List<HudiColumnHandle> partitionColumnHandles, String partitionPathValue, String partitionName)
```

Comment:
nit: rename to `evaluatePartitionPredicate`?

Reply:
> Thank you for raising this detailed PR. My other PR based on RFC-58 was trying to do data skipping using metadata table in a more generic way. Basically rather than introducing the changes in specific query engines like presto and trino, the idea was to introduce the changes as part of hudi itself and simply call them from presto/trino. Anyways I can make changes accordingly later. I am still going through the changes and have given few comments for changes/clarification. Please have a look.

yes  If rfc-58 is completed, we only need to convert the presto filter into a hudi filter, and then call the interface directly, just like iceberg. and also RFC-64 is abstracting interfaces, however this may take a long time.

once rfc-58/rfc-64 completed,  we can remove those logical directly.

Reply:
Exactly, I am aligned on this.